### PR TITLE
Fix http3 servers dying after reload

### PR DESCRIPTION
**Pull request by renbou**
This PR should finally resolve issues with http3 servers dying after Caddy's reload. Currently it pins the quic-go dependency to the commit in [lucas-clemente/quic-go](https://github.com/lucas-clemente/quic-go) where the necessary changes are made. I've tested everything out locally and used pprof to check that no goroutines are being leaked.

- Added caddy.ListenQUIC which returns a QUIC listener valid
for reloading in Caddy (for example reloading the http3 server).
- Finally resolved issues encountered in caddyserver/caddy#2727

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -16,7 +16,7 @@ require (
 	github.com/google/uuid v1.3.0
 	github.com/klauspost/compress v1.15.0
 	github.com/klauspost/cpuid/v2 v2.0.11
-	github.com/lucas-clemente/quic-go v0.25.0
+	github.com/lucas-clemente/quic-go v0.25.1-0.20220321092029-b7e93b54c9cc
```
Comment (marten-seemann) [inline go.mod:19]: Do you guys want a release? Just let me know, I can cut one.

Comment (renbou) [reply go.mod:19]: Let's do one after https://github.com/lucas-clemente/quic-go/issues/3350 :D

Comment (JeDaYoshi) [reply go.mod:19]: Can we update this now that #3352 was merged? :eyes:

Comment (francislavoie) [reply go.mod:19]: I think you meant https://github.com/lucas-clemente/quic-go/issues/3352, but yep, I'm sure @renbou will be on it as soon as he can!

Comment (JeDaYoshi) [reply go.mod:19]: Yep, my bad - thanks!

Comment (marten-seemann) [reply go.mod:19]: Cutting the release will be my responsibility - but please give CI enough time to run before asking for the release ;)
This PR was literally merge less than 10 minutes ago.

Comment (JeDaYoshi) [reply go.mod:19]: All good! Didn't want to sound impatient.

**Review by mholt — APPROVED** [review]
If this actually works, this is awesome. Thanks so much @renbou -- can't wait to try it out! (The code looks good though.)

File: go.mod (hunk 1 above)
Comment (marten-seemann) [reply go.mod:19]: v0.26.0 was shipped: https://github.com/lucas-clemente/quic-go/releases/tag/v0.26.0

Comment (JeDaYoshi) [reply go.mod:19]: Thank you!

**Review by JeDaYoshi — APPROVED** [review]

**Comment by renbou** [comment]
Thank you @marten-seemann! Might as well merge in the optimization in the same PR.

**Review by francislavoie — APPROVED** [review]

**Pull request merged**